File: TopKFrequentElements/TopKFrequentElements.hs
import Data.List qualified as L
import Data.Map qualified as M

topKFrequentElements :: [Int] -> Int -> [Int]
topKFrequentElements nums k = take k $ fst <$> sortedFreqs
  where
    sortedFreqs = L.sortBy (\(_, f1) (_, f2) -> compare f2 f1) freqs
    freqs = M.toList $ foldr go M.empty nums
    go :: Int -> M.Map Int Int -> M.Map Int Int
    go n = M.insertWith (+) n 1

replace :: [a] -> Int -> a -> [a]
replace as n a = a0 ++ a : af
  where
    (a0, _ : af) = L.splitAt n as

topKFrequentElements2 :: [Int] -> Int -> [Int]
topKFrequentElements2 nums k = take k $ concat freq
  where
    freq = foldr go [[] | _ <- [1 .. len + 1]] (M.toList count)
    go :: (Int, Int) -> [[Int]] -> [[Int]]
    go (n, f) ns = replace ns (len - f) (ns !! (len - f) ++ [n])
    len = length nums
    count = foldr go' M.empty nums
    go' :: Int -> M.Map Int Int -> M.Map Int Int
    go' n = M.insertWith (+) n 1

File: TopKFrequentElements/top_k_frequent_elements.py
def top_k_frequent_elements(nums: [int], k: int) -> [int]:
    from collections import defaultdict

    freqs = defaultdict(int)
    for n in nums:
        freqs[n] += 1
    ordered_freqs = sorted(freqs.items(), key=lambda p: p[1], reverse=True)
    ans = [ordered_freqs[i][0] for i in range(k)]
    return ans

def top_k_frequent_elements2(nums: [int], k: int) -> [int]:
    from collections import defaultdict

    count = defaultdict(int)
    freq = ([] for i in range(len(nums) + 1))
    for n in nums:
        count[n] += 1
    for n, c in count.items():
        freq[c].append(n)
    ans = []
    c = 0
    for ns in freq[::-1]:
        for n in ns:
            ans.append(n)
            c += 1
            if c == k:
                return ans
